src/x86/intel_joule_expansion.c

```diff
@@ -145,6 +181,115 @@ mraa_joule_expansion_board()
     b->spi_bus[2].bus_id = 32766;
     b->spi_bus[2].slave_s = 1;
 
+    // Aio shield detection
+    // Only if i2c_designware.0 was successfully found
+    if (i2c_aio_bus != -1) {
+
+        syslog(LOG_NOTICE, "Attempting shield autodetection on I2C bus %d...", i2c_aio_bus);
+
+        i2c_smbus_ioctl_data_t sm;
+        i2c_smbus_data_t d;
+        int fp;
+
+        char i2c_fh[64];
+        snprintf(i2c_fh, 64, "/dev/i2c-%d", i2c_aio_bus);
+
+        sm.read_write = I2C_SMBUS_READ;
+        sm.command = 0x01;
+        sm.size = I2C_SMBUS_WORD_DATA; // LSB first
+        sm.data = &d;
+
+        if ((fp = open(i2c_fh, O_RDWR)) < 1)
+            syslog(LOG_ERR, "Failed to open I2C bus: %s", strerror(errno));
+        if (ioctl(fp, I2C_SLAVE_FORCE, 0x49) < 0)
+            syslog(LOG_ERR, "Failed to set I2C address: %s", strerror(errno));
+        if (ioctl(fp, I2C_SMBUS, &sm) < 0) {
+            syslog(LOG_NOTICE, "No device at I2C address 0x49");
+
+            // no DFRobot shield, try Grove
+            if (ioctl(fp, I2C_SLAVE_FORCE, 0x48) < 0)
+                syslog(LOG_ERR, "Failed to set I2C address: %s", strerror(errno));
+            if (ioctl(fp, I2C_SMBUS, &sm) < 0)
+                syslog(LOG_NOTICE, "No device at I2C address 0x48");
+                // no shield
+            else {
+                // load ads1015 at 0x48
+                syslog(LOG_NOTICE, "Loading ti-ads1015 module for Grove Shield");
+                if (system("modprobe ti-ads1015") < 0) {
+                    syslog(LOG_NOTICE, "Failed, are you running the latest Joule kernel?");
+                } else {
+                    char command[128];
+                    snprintf(command, 128, "echo ads1015 0x48 >/sys/bus/i2c/devices/i2c-%d/new_device", i2c_aio_bus);
+                    if (system(command) < 0) {
+                        syslog(LOG_ERR, "Failed to add ads1015 device");
+                    } else {
+                        // success, setup aio pins
+                        b->aio_count = 4;
+                        b->adc_raw = 11; // 12-bit ads1015 - sign bit
+                        b->adc_supported = 10;
+                        pincount += b->aio_count;
+                        shield = GROVE;
+
+                        // max sample rate and 6.144 V reference
+                        // additional aio functions can be added to make up for loss of precision at 3V3, 5V, etc
+                        for (int i = 0; i < b->aio_count; i++) {
+                            snprintf(command, 128, "echo 3300 >/sys/bus/iio/devices/iio:device0/in_voltage%d_sampling_frequency", i);
+                            system(command);
+                            snprintf(command, 128, "echo 3 >/sys/bus/iio/devices/iio:device0/in_voltage%d_scale", i);
```

Comment:
Again random system() calls should not be something we do. No reason we can't do this with the Aio class?